Lib/test/test_tabnanny.py

```diff
@@ -0,0 +1,356 @@
+
+    def test_correct_directory_verbose(self):
+        """Directory containing few error free python source code files.
+
+        Because order of files returned by `os.lsdir()` is not fixed, verify the
+        existence of each output lines at `stdout` using `in` operator.
+        `verbose` mode of `tabnanny.verbose` asserts `stdout`.
+        """
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            lines = ["%r: listing directory\n" % (tmp_dir,),]
+            with TemporaryPyFile(
+                SOURCE_CODES["error_free"], directory=tmp_dir
+            ) as file1_path, TemporaryPyFile(
+                SOURCE_CODES["error_free"], directory=tmp_dir
+            ) as file2_path:
+                for file_path in (file1_path, file2_path):
+                    lines.append(
+                        "%r: Clean bill of health.\n" % (file_path,)
+                    )
+
+                tabnanny.verbose = 1
+                with captured_stdout() as stdout, captured_stderr() as stderr:
+                    tabnanny.check(tmp_dir)
+                stdout = stdout.getvalue()
+                for line in lines:
+                    self.assertIn(line, stdout)
+                self.assertEqual(stderr.getvalue(), "")
+
+    def test_correct_directory(self):
+        """Directory which contains few error free python source code files."""
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            with TemporaryPyFile(
+                SOURCE_CODES["error_free"], directory=tmp_dir
+            ) as file_path:
+                self.verify_tabnanny_check(tmp_dir)
+
+    def test_when_wrong_indented(self):
+        """A python source code file eligible for raising `IndentationError`."""
+        with TemporaryPyFile(
+            SOURCE_CODES["wrong_indented"]
+        ) as file_path:
+            err = ('unindent does not match any outer indentation level'
+                ' (<tokenize>, line 3)\n')
+            err = "%r: Indentation Error: %s" % (file_path, err)
+            self.verify_tabnanny_check(file_path, err=err)
+
+    def test_when_tokenize_tokenerror(self):
+        """A python source code file eligible for raising 'tokenize.TokenError'.
+        """
+        with TemporaryPyFile(
+            SOURCE_CODES["incomplete_expression"]
+        ) as file_path:
+            err = "('EOF in multi-line statement', (7, 0))\n"
+            err = "%r: Token Error: %s" % (file_path, err)
+            self.verify_tabnanny_check(file_path, err=err)
+
+    def test_when_nannynag_error_verbose(self):
+        """A python source code file eligible for raising `tabnanny.NannyNag`.
+
+        Tests will assert `stdout` after activating `tabnanny.verbose` mode.
+        """
+        with TemporaryPyFile(
+            SOURCE_CODES["nannynag_errored"]
+        ) as file_path:
+            out = "%r: *** Line 3: trouble in tab city! ***\n" % (file_path,)
+            out += "offending line: '\\tprint(\"world\")\\n'\n"
```

Comment:
Don't worry too much about the exact output being formatted as it is. For instance, if someone tweaked the grammar of the sentence the test should still pass. Do still check for semantically important info, though, like that the offending line is in the output.

Reply:
Maybe the test is too strict, but IMHO we have been too pendantic on this nice PR which has been proposed 3 months ago. tabnanny has currently no test. If the test fails tomorrow, we will just fix the test. No need to write the perfect test here.